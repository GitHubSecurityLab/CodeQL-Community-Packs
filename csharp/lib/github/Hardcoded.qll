import csharp

abstract class HardcodedValues extends DataFlow::ExprNode { }

class NonEmptyStringLiteral extends HardcodedValues {
  NonEmptyStringLiteral() { this.getExpr().(StringLiteral).getValue().length() > 1 }
}

class ByteArrayLiteral extends HardcodedValues {
  ByteArrayLiteral() {
    this.getExpr() =
      any(ArrayCreation ac |
        ac.getArrayType().getElementType() instanceof ByteType and
        ac.hasInitializer()
      )
  }
}

class CharArrayLiteral extends HardcodedValues {
  CharArrayLiteral() {
    this.getExpr() =
      any(ArrayCreation ac |
        ac.getArrayType().getElementType() instanceof CharType and
        ac.hasInitializer()
      )
  }
}
