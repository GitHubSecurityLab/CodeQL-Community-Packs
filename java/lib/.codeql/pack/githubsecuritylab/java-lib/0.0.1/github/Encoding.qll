import java
import semmle.code.java.dataflow.DataFlow
import semmle.code.java.dataflow.FlowSources
import semmle.code.java.dataflow.TaintTracking2

module Base64 {
  abstract class Encoding extends DataFlow::Node { }

  abstract class Decoding extends DataFlow::Node { }

  // codeql/java/ql/lib/semmle/code/java/security/HardcodedCredentialsApiCallQuery.qll
  class Encoders extends Base64::Encoding {
    Encoders() {
      exists(MethodAccess ma |
        ma.getMethod()
            .hasQualifiedName([
                "java.util", "cn.hutool.core.codec", "org.apache.shiro.codec",
                "apache.commons.codec.binary", "org.springframework.util"
              ], ["Base64$Encoder", "Base64", "Base64Utils"],
              [
                "encode", "encodeToString", "encodeBase64", "encodeBase64Chunked",
                "encodeBase64String", "encodeBase64URLSafe", "encodeBase64URLSafeString"
              ])
      |
        this.asExpr() = ma.getArgument(0)
      )
    }
  }

  class Decoders extends Base64::Decoding {
    Decoders() {
      exists(MethodAccess ma |
        ma.getMethod()
            .hasQualifiedName([
                "java.util", "cn.hutool.core.codec", "org.apache.shiro.codec",
                "apache.commons.codec.binary", "org.springframework.util"
              ], ["Base64$Decoder", "Base64", "Base64Utils"], ["decode", "decodeBase64"])
      |
        this.asExpr() = ma.getArgument(0)
      )
    }
  }
}
