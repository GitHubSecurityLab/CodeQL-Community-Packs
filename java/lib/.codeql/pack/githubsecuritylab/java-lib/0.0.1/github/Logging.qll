import java
import semmle.code.java.dataflow.DataFlow
import semmle.code.java.dataflow.FlowSources
import semmle.code.java.dataflow.TaintTracking2

abstract class LoggingMethodsSinks extends DataFlow::Node { }

// TODO: Use the exists libs in CodeQL to extend this
class PrintMethods extends LoggingMethodsSinks {
  PrintMethods() {
    exists(MethodAccess ma |
      ma.getMethod().getDeclaringType().hasQualifiedName("java.io", _) and
      (
        ma.getMethod().hasName("println") or
        ma.getMethod().hasName("print") or
        ma.getMethod().getName() = "append" or
        ma.getMethod().getName() = "format" or
        ma.getMethod().getName() = "write"
      ) and
      this.asExpr() = ma.getArgument(0)
    )
  }
}

// Ref :: https://github.com/github/codeql/blob/main/java/ql/src/experimental/Security/CWE/CWE-532/SensitiveInfoLog.ql
// Ref :: https://github.com/github/codeql/blob/main/java/ql/src/experimental/semmle/code/java/Logging.qll#L34
abstract class LoggerType extends RefType { }

class StandardLoggersType extends LoggerType {
  StandardLoggersType() {
    // JBoss Logging (`org.jboss.logging.Logger` in some implementations like JBoss Application Server 4.0.4 did not implement `BasicLogger`)
    this.hasQualifiedName("org.jboss.logging", ["BasicLogger", "Logger"]) or
    this.hasQualifiedName("org.slf4j.spi", "LoggingEventBuilder") or
    this.hasQualifiedName("org.slf4j", "Logger") or
    this.hasQualifiedName("org.scijava.log", "Logger") or
    this.hasQualifiedName("com.google.common.flogger", "LoggingApi") or
    this.hasQualifiedName("java.lang", "System$Logger") or
    this.hasQualifiedName("java.util.logging", "Logger") or
    this.hasQualifiedName("android.util", "Log")
  }
}

class Log4jLoggerType extends LoggerType {
  Log4jLoggerType() {
    this.hasQualifiedName("org.apache.log4j", "Category") or // Log4j 1
    this.hasQualifiedName("org.apache.logging.log4j", ["Logger", "LogBuilder"]) or // Log4j 2
    this.hasQualifiedName("org.apache.commons.logging", "Log")
  }
}

class LoggingMethods extends LoggingMethodsSinks {
  LoggingMethods() {
    exists(MethodAccess ma |
      ma.getMethod().getDeclaringType() instanceof LoggerType and
      (
        ma.getMethod().hasName("debug") or
        ma.getMethod().hasName("trace") or
        ma.getMethod().hasName("debugf") or
        ma.getMethod().hasName("debugv")
      ) and //Check low priority log levels which are more likely to be real issues to reduce false positives
      this.asExpr() = ma.getAnArgument()
    )
  }
}
