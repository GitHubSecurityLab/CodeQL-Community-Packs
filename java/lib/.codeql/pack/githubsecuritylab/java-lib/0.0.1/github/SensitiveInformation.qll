import java
import semmle.code.java.dataflow.DataFlow
import semmle.code.java.dataflow.FlowSources
import semmle.code.java.dataflow.TaintTracking2

abstract class SensitiveInformationSources extends DataFlow::Node { }

class HttpSession extends SensitiveInformationSources {
  HttpSession() {
    exists(MethodAccess ma |
      // https://docs.oracle.com/javaee/5/api/javax/servlet/http/HttpSession.html
      // Assumption: Nothing from the Session object should be logged
      ma.getMethod().getDeclaringType().hasQualifiedName("javax.servlet.http", "HttpSession") and
      this.asExpr() = ma
    )
  }
}

class Properties extends SensitiveInformationSources {
  Properties() {
    exists(MethodAccess ma |
      ma.getMethod().hasName("getProperty") and
      this.asExpr() = ma
    )
  }
}

class SensitiveVariables extends SensitiveInformationSources {
  SensitiveVariables() {
    exists(Variable v |
      (
        // User data
        v.getName().toLowerCase().regexpMatch(".*(username|passport|fingerprint|dob|ssi).*")
        or
        // Creds / Secrets / Tokens
        v.getName().toLowerCase().regexpMatch(".*(password|pwd|hash|secret|token|session).*")
        or
        // Card Numbers
        v.getName().toLowerCase().regexpMatch(".*(cardnumber|cvv|sortcode|accountnumber).*")
      ) and
      this.asExpr() = v.getAnAccess()
    )
  }
}
