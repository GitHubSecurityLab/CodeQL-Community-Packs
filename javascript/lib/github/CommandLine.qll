import semmle.javascript.dataflow.DataFlow

class CommandLineArgument extends DataFlow::Node {
    CommandLineArgument() {
        this = DataFlow::globalVarRef("process").getAPropertyRead("argv").getAPropertyReference()
    }
}
