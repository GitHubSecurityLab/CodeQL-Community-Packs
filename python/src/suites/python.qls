- description: "GitHub's Community Packs Python Extended Suite"

- queries: '.'
  from: githubsecuritylab/codeql-python-queries

- include:
    kind:
    - problem
    - path-problem
    precision:
    - very-high
    - high
    - medium
    - low

# Remove debugging, and audit queries
- exclude:
    tags contain:
      - debugging
      - audit

# Remove local testing folders
- exclude:
    query path:
      - /testing\/.*/
