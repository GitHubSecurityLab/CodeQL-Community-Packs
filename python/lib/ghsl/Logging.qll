import python
import semmle.python.ApiGraphs

abstract class LoggingSinks extends DataFlow::Node { }

class PrintMethod extends LoggingSinks {
  PrintMethod() {
    exists(DataFlow::Node call |
      call = API::builtin("print").getACall() and
      call.getScope().inSource() and
      this = call
    )
  }
}

class LoggingFramework extends LoggingSinks {
  LoggingFramework() {
    exists(DataFlow::Node call, API::Node node |
      (
        (
          node = API::moduleImport("logging")
          or
          node = API::moduleImport("logging").getMember("getLogger").getReturn()
        ) and
        call = node.getMember(["info", "debug", "warning", "error"]).getACall()
      ) and
      call.getScope().inSource() and
      this = call
    )
  }
}
