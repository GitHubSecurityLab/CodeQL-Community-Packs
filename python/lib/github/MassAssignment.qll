import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.dataflow.new.TaintTracking
import semmle.python.Concepts
import semmle.python.dataflow.new.RemoteFlowSources
import semmle.python.dataflow.new.BarrierGuards
import semmle.python.ApiGraphs
import github.LocalSources

module MassAssignment {
  abstract class Sources extends DataFlow::Node { }

  abstract class Sinks extends DataFlow::Node { }

  abstract class Sanitizer extends DataFlow::Node { }

  // Setattr build into Python
  class Setattr extends MassAssignment::Sinks {
    Setattr() {
      (
        // > `setattr(obj, SINK, value)`
        this = API::builtin("setattr").getACall().getArg(1)
        or
        // > __setattr__(SINK, value)
        exists(Value value, CallNode call |
          value.getName() = "__setattr__" and
          call = value.getACall() and
          this.asCfgNode() = call.getArg(0)
        )
      ) and
      this.getScope().inSource()
    }
  }

  class MassAssignmentLocalConfig extends TaintTracking::Configuration {
    MassAssignmentLocalConfig() { this = "Mass Assignment Config" }
  
    override predicate isSource(DataFlow::Node source) { source instanceof LocalSources::Range }
  
    override predicate isSink(DataFlow::Node sink) { sink instanceof MassAssignment::Sinks }
  
    override predicate isSanitizer(DataFlow::Node node) { node instanceof MassAssignment::Sanitizer }
  }
  
  class MassAssignmentConfig extends TaintTracking::Configuration {
    MassAssignmentConfig() { this = "Mass Assignment Config" }
  
    override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource::Range }
  
    override predicate isSink(DataFlow::Node sink) { sink instanceof MassAssignment::Sinks }
  
    override predicate isSanitizer(DataFlow::Node node) { node instanceof MassAssignment::Sanitizer }
  }
  
}
