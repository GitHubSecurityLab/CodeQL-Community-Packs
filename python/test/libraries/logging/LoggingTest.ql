import python
import github.Logging

from LoggingSinks sources
select sources
