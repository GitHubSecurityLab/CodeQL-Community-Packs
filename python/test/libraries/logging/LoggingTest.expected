| app.py:6:1:6:17 | ControlFlowNode for print() |
| app.py:12:1:12:39 | ControlFlowNode for Attribute() |
| app.py:13:1:13:40 | ControlFlowNode for Attribute() |
| app.py:14:1:14:44 | ControlFlowNode for Attribute() |
| app.py:15:1:15:41 | ControlFlowNode for Attribute() |
| app.py:19:1:19:38 | ControlFlowNode for Attribute() |
| app.py:20:1:20:39 | ControlFlowNode for Attribute() |
| app.py:21:1:21:43 | ControlFlowNode for Attribute() |
| app.py:22:1:22:40 | ControlFlowNode for Attribute() |
